### DISCUSSION TITLE
Tailwind not working in create-react-app

### DISCUSSION THREAD

**Discussion by erpriliano** [tailwindlabs/tailwindcss#2833] [Help] [answered]
Hi there,

Basically I've already followed the steps to include tailwindcss into my react-app, these are the steps I followed :

1. Create react-app from npx using `npx create-react-app [project-name]` and I open it
2. Then I installed tailwindcss using this command `npm install tailwindcss autoprefixer postcss-cli`
3. I created config file and postcss.config.js using these commands `npx tailwindcss init` and `touch postcss.config.js`. I'm doing both in my root project folder 
4. Inside my **src** folder, I created folder called **styles** with two files (tailwind.css & main.css)
5. Then I edited my script inside my package.json

**My postcss.config.js**
```js
const tailwindcss = require("tailwindcss");
module.exports = {
    plugins: [
        tailwindcss("./tailwind.config.js"),
        require("autoprefixer")
    ]
};
```
**Inside postcss.config.js I keep getting a warning says:**
> Could not find a declaration file for module 'tailwindcss'. '/Users/erpriliano/React/kajol-web/node_modules/tailwindcss/lib/index.js' implicitly has an 'any' type.
  Try `npm install @types/tailwindcss` if it exists or add a new declaration (.d.ts) file containing `declare module 'tailwindcss';`

**My tailwind.css**
```css
@tailwind base;
@tailwind components;
@tailwind utilities;
```

**My package.json**
```json
{
  "name": "kajol-web",
  "version": "0.1.0",
  "private": true,
  "dependencies": {
    "@testing-library/jest-dom": "^5.11.4",
    "@testing-library/react": "^11.1.0",
    "@testing-library/user-event": "^12.1.10",
    "autoprefixer": "^9.0.0",
    "postcss-cli": "^7.1.2",
    "react": "^17.0.1",
    "react-dom": "^17.0.1",
    "react-scripts": "4.0.0",
    "tailwindcss": "^2.0.1",
    "web-vitals": "^0.2.4"
  },
  "scripts": {
    "start": "npm run watch:css && react-scripts start",
    "build": "npm run build:css && react-scripts build",
    "test": "react-scripts test",
    "eject": "react-scripts eject",
    "build:css": "postcss src/styles/tailwind.css -o src/styles/main.css",
    "watch:css": "postcss src/styles/tailwind.css -o src/styles/main.css"
  },
  "eslintConfig": {
    "extends": [
      "react-app",
      "react-app/jest"
    ]
  },
  "browserslist": {
    "production": [
      ">0.2%",
      "not dead",
      "not op_mini all"
    ],
    "development": [
      "last 1 chrome version",
      "last 1 firefox version",
      "last 1 safari version"
    ]
  }
}
```

Then I ran `npm start` but always got these errors: 

> kajol-web@0.1.0 start /Users/erpriliano/React/kajol-web
> npm run watch:css && react-scripts start

> kajol-web@0.1.0 watch:css /Users/erpriliano/React/kajol-web
> postcss src/styles/tailwind.css -o src/styles/main.css

> Error: PostCSS plugin postcss-nested requires PostCSS 8.
> Migration guide for end-users:
> https://github.com/postcss/postcss/wiki/PostCSS-8-for-end-users
>    at Processor.normalize (/Users/erpriliano/React/kajol-web/node_modules/postcss/lib/processor.js:167:15)
>    at new Processor (/Users/erpriliano/React/kajol-web/node_modules/postcss/lib/processor.js:56:25)
>    at postcss (/Users/erpriliano/React/kajol-web/node_modules/postcss/lib/postcss.js:55:10)
>    at /Users/erpriliano/React/kajol-web/node_modules/tailwindcss/lib/util/parseObjectStyles.js:24:33
>    at arrayMap (/Users/erpriliano/React/kajol-web/node_modules/lodash/lodash.js:639:23)
>    at map (/Users/erpriliano/React/kajol-web/node_modules/lodash/lodash.js:9580:14)
>   at Function.flatMap (/Users/erpriliano/React/kajol-web/node_modules/lodash/lodash.js:9283:26)
>    at parseObjectStyles (/Users/erpriliano/React/kajol-web/node_modules/tailwindcss/lib/util/parseObjectStyles.js:23:26)
>    at parseObjectStyles (/Users/erpriliano/React/kajol-web/node_modules/tailwindcss/lib/util/parseObjectStyles.js:20:12)
>    at /Users/erpriliano/React/kajol-web/node_modules/tailwindcss/lib/util/processPlugins.js:37:123
> npm ERR! code ELIFECYCLE
> npm ERR! errno 1
> npm ERR! kajol-web@0.1.0 watch:css: `postcss src/styles/tailwind.css -o src/styles/main.css`
> npm ERR! Exit status 1
> npm ERR! 
> npm ERR! Failed at the kajol-web@0.1.0 watch:css script.
> npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

> npm ERR! A complete log of this run can be found in:
> npm ERR!     /Users/erpriliano/.npm/_logs/2020-11-20T00_38_53_408Z-debug.log
> npm ERR! code ELIFECYCLE
> npm ERR! errno 1
> npm ERR! kajol-web@0.1.0 start: `npm run watch:css && react-scripts start`
> npm ERR! Exit status 1
> npm ERR! 
> npm ERR! Failed at the kajol-web@0.1.0 start script.
> npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

> npm ERR! A complete log of this run can be found in:
> npm ERR!     /Users/erpriliano/.npm/_logs/2020-11-20T00_38_53_429Z-debug.log

I googled and I found the same problem in stackoverflow as [this](https://stackoverflow.com/questions/63907516/attempt-to-launch-react-app-with-tailwind-fails) and it already two months ago but there's no satisfying answer of how to fix it. Could you please help us with this problem?

**Comment by adamwathan** [comment, marked as answer]
Running `npm install postcss@latest` should fix it 👍🏻

**Reply by erpriliano** [reply]
Yes, problem solved and I'm realised by now that I haven't installed postcss at all in that package.json. Thanks :)

**Reply by dwbelliston** [reply]
@erpriliano - in the tailwind docs its mentioned using CRACO to get it working, but here you are not, can you help me learn why this way instead of via CRACO?

**Reply by erpriliano** [reply]
To be honest with you, the moment I asked the question the docs for installing tailwind in CRA haven't published yet so I've to add it manually by using previous docs and need to did some improvement a bit as needed in my project.

**Reply by dwbelliston** [reply]
ah! okay - gotcha - that makes sense. thanks for sharing - was making sure i wasnt missing something.

**Reply by max-programming** [reply]
Thank you so much for helping 🤍

**Reply by Tasha53505** [reply]
Fixed now for me too with "npm install postcss@latest"

24th June 2022  !

**Comment by phsamudev** [comment]
I've managed to get it to work by following the guide for **PostCSS 7 compatibility build** [here](https://tailwindcss.com/docs/installation#post-css-7-compatibility-build). I just changed it to `yarn add tailwindcss@npm:@tailwindcss/postcss7-compat postcss-cli@^7 autoprefixer@^9` since I'm using **yarn** and **postcss-cli**.

**Reply by felipe-campanhol** [reply]
Thanks!, installing this way solved my issue

**Reply by tfrank11** [reply]
Thanks

**Comment by Asaju5** [comment]
I have tried every possible way to install tailwindcss in my createReactApp project but never worked, please I need some help

**Reply by erpriliano** [reply]
Now that tailwind has provided us with some comprehensive explanation of how to install it in our CRA project. You might want to see [this](https://tailwindcss.com/docs/guides/create-react-app)

**Comment by Asaju5** [comment]
I followed every single step in the installation page, but I keep getting
an error message, do you think something might be wrong with my pc?

**Reply by erpriliano** [reply]
Can you tell us what's the error? And also explain to us of how you install tailwind to your project step by step

**Comment by Asaju5** [comment]
I followed all the steps in *tailwindcss.com/docs/guides/create-react-app,
and even watched a video on YouTube:

Here’s the error message:-



SyntaxError: Invalid or unexpected token
    at wrapSafe (internal/modules/cjs/loader.js:979:16)
    at Module._compile (internal/modules/cjs/loader.js:1027:27)
    at Object.Module._extensions..js
(internal/modules/cjs/loader.js:1092:10)
    at Module.load (internal/modules/cjs/loader.js:928:32)
    at Function.Module._load (internal/modules/cjs/loader.js:769:14)
    at Module.require (internal/modules/cjs/loader.js:952:19)
    at require (internal/modules/cjs/helpers.js:88:18)
    at Object.<anonymous> (C:\Users\My
Pc\Documents\tailwind-app\node_modules\tailwindcss\lib\util\generateVariantFunction.js:14:16)
    at Module._compile (internal/modules/cjs/loader.js:1063:30)
    at Object.Module._extensions..js
(internal/modules/cjs/loader.js:1092:10)
(Use `node --trace-warnings ...` to show where the warning was created)
(node:1100) UnhandledPromiseRejectionWarning: Unhandled promise rejection.
This error originated either by throwing inside of an async function
without a catch block, or by rejecting a promise which was not handled with
.catch(). To terminate the node process on unhandled promise rejection, use
the CLI flag `--unhandled-rejections=strict` (see
https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection
id: 2)
(node:1100) [DEP0018] DeprecationWarning: Unhandled promise rejections are
deprecated. In the future, promise rejections that are not handled will
terminate the Node.js process with
a non-zero exit code.

**Comment by adamwathan** [comment]
@Asaju5 What version of Node are you using? Tailwind requires a minimum of v12 in case you are on an earlier version.

**Reply by JoseAlz** [reply]
work! ajaja Setting up Tailwind CSS
Tailwind CSS requires Node.js 12.13.0 or higher.

**Comment by Asaju5** [comment]
I upgraded my node to version 14.15.3 LTS yesterday, honestly this is so
frustrating, tailwind is exactly the framework I’ve been looking for

>

**Reply by erpriliano** [reply]
I'm trying to guess that your problem is with your Node, will you try to re-install node on your PC? See if that works..

**Reply by smrnjeet222** [reply]
@Asaju5 I'm also facing the exact same issue

**Reply by sagor9116** [reply]
The new documentation is working fine. https://tailwindcss.com/docs/guides/create-react-app . go to the link and try to the installation part step by step. I hope it will work

**Reply by JoseAlz** [reply]
ailwind-app@0.1.0 start /home/phi/React_projectos/tailwind-app
> craco start

(node:23947) UnhandledPromiseRejectionWarning: TypeError: Object.fromEntries is not a function
    at Object.<anonymous> (/home/phi/React_projectos/tailwind-app/node_modules/tailwindcss/lib/jit/corePlugins.js:242:13)
    at Module._compile (internal/modules/cjs/loader.js:778:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:789:10)
    at Module.load (internal/modules/cjs/loader.js:653:32)
    at tryModuleLoad (internal/modules/cjs/loader.js:593:12)
    at Function.Module._load (internal/modules/cjs/loader.js:585:3)
    at Module.require (internal/modules/cjs/loader.js:692:17)
    at require (internal/modules/cjs/helpers.js:25:18)
    at Object.<anonymous> (/home/phi/React_projectos/tailwind-app/node_modules/tailwindcss/lib/jit/lib/setupContextUtils.js:36:43)
    at Module._compile (internal/modules/cjs/loader.js:778:30)
(node:23947) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). (rejection id: 2)
(node:23947) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.

**Comment by ogcsantos** [comment]
Hello, I'm having problems starting my React project with TailwindCss with both NPM and YARN represent the same error, and I'm definitely not sure what to do to solve this problem. (Obs: I'm not fluent in the English language, I'm Brazilian, I'm sorry for any mistakes in transcription.

>  yarn start
yarn run v1.22.5
$ craco start
(node:31360) UnhandledPromiseRejectionWarning: TypeError [ERR_INVALID_ARG_VALUE]: The argument 'id' must be a non-empty string. Received ''
    at Module.require (internal/modules/cjs/loader.js:1020:11)
    at require (internal/modules/cjs/helpers.js:72:18)
    at getConfigAsObject (C:\_xampp_7.4\htdocs\_services\2k21\tchaau.com.br\frontend\node_modules\@craco\craco\lib\config.js:46:20)
    at loadCracoConfigAsync (C:\_xampp_7.4\htdocs\_services\2k21\tchaau.com.br\frontend\node_modules\@craco\craco\lib\config.js:69:34)
    at Object.<anonymous> (C:\_xampp_7.4\htdocs\_services\2k21\tchaau.com.br\frontend\node_modules\@craco\craco\scripts\start.js:22:1)
    at Module._compile (internal/modules/cjs/loader.js:1137:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1157:10)
    at Module.load (internal/modules/cjs/loader.js:985:32)
    at Function.Module._load (internal/modules/cjs/loader.js:878:14)
    at Function.executeUserEntryPoint [as runMain] (internal/modules/run_main.js:71:12)
(node:31360) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). To terminate the node process on unhandled promise rejection, use the CLI flag `--unhandled-rejections=strict` (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 2)      
(node:31360) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.)

**Reply by ogcsantos** [reply]
![image](https://user-images.githubusercontent.com/39639379/107828619-fd14f400-6d67-11eb-8270-3ed500ac49b1.png)
![image](https://user-images.githubusercontent.com/39639379/107828668-174ed200-6d68-11eb-910e-68ef4500d6d8.png)

**Reply by wilcus** [reply]
Probably you do not put the configuration files in the root folder

**Reply by TutorialDoctor** [reply]
> Probably you do not put the configuration files in the root folder

This was my issue.

**Comment by JimLynchCodes** [comment]
honestly, the install process is insanely complicated, and there are numerous conflicting guides on the internet. 

This is so horrible... smh

**Reply by TobiLight** [reply]
This!

**Comment by max-programming** [comment]
I would suggest using [Vite](https://vitejs.dev) instead of Create React App.

[Creating a Vite app](https://vitejs.dev/guide/#scaffolding-your-first-vite-project)

- It is the fastest (give it a try and you'll realize)
- Tailwind installation is damn easy

**Comment by TobiLight** [comment]
I followed every step here and it works in development. But when I deploy to Heroku, the error log says:

Error: Cannot find module tailwindcss 🤷🏽‍♂️

**Reply by davidofug** [reply]
For me on Development it gives me the error.

**Comment by DavidSalemLee** [comment]
I have the same error. Works perfectly in development but in Heroku I am getting Error: cannot find module 'tailwindcss' and the build doesn't complete

**Comment by sammiearchie77** [comment]
Hi everyone!
Please I have a problem with tailwindcss using react.
My styling is not working on production stage but on development everything works fine.
What could be the problem??

this is my package.json
`{
  "name": "cra-pwa",
  "version": "0.1.0",
  "private": true,
  "homepage": "https://sammiearchie77.github.io/my-portfolio",
  "dependencies": {
    "@craco/craco": "^6.2.0",
    "@heroicons/react": "^1.0.3",
    "@testing-library/jest-dom": "^5.14.1",
    "@testing-library/react": "^11.2.7",
    "@testing-library/user-event": "^12.8.3",
    "axios": "^0.21.1",
    "gh-pages": "^3.2.3",
    "react": "^17.0.2",
    "react-dom": "^17.0.2",
    "react-scripts": "4.0.3",
    "web-vitals": "^0.2.4",
    "workbox-background-sync": "^5.1.4",
    "workbox-broadcast-update": "^5.1.4",
    "workbox-cacheable-response": "^5.1.4",
    "workbox-core": "^5.1.4",
    "workbox-expiration": "^5.1.4",
    "workbox-google-analytics": "^5.1.4",
    "workbox-navigation-preload": "^5.1.4",
    "workbox-precaching": "^5.1.4",
    "workbox-range-requests": "^5.1.4",
    "workbox-routing": "^5.1.4",
    "workbox-strategies": "^5.1.4",
    "workbox-streams": "^5.1.4"
  },
  "scripts": {
    "start": "craco start",
    "predeploy": "npm run build",
    "deploy": "gh-pages -d build",
    "build": "craco build",
    "test": "craco test",
    "eject": "react-scripts eject"
  },
  "eslintConfig": {
    "extends": [
      "react-app",
      "react-app/jest"
    ]
  },
  "browserslist": {
    "production": [
      ">0.2%",
      "not dead",
      "not op_mini all"
    ],
    "development": [
      "last 1 chrome version",
      "last 1 firefox version",
      "last 1 safari version"
    ]
  },
  "devDependencies": {
    "autoprefixer": "^9.8.6",
    "postcss": "^7.0.36",
    "tailwindcss": "npm:@tailwindcss/postcss7-compat@^2.2.7"
  }
}
`

this is my tailwind.config.js 
`module.exports = {
  purge: ['./src/**/*.{js,jsx,ts,tsx', './public/index.html'],
  darkMode: false, // or 'media' or 'class'
  theme: {
    extend: {},
  },
  variants: {
    extend: {},
  },
  plugins: [],
}
`

**Comment by sammiearchie77** [comment]
I did. However, since I didn't see anyone to solve it, I ended up using the
one I'm proficient in.

**Reply by rahulrg1** [reply]
if you have done whatever is stated here  " [https://tailwindcss.com/docs/guides/create-react-app](url) "
after this if everything is working fine except your tailwind css then try one way to fix (if you haven't tried it yet)
1. go to your main folder path in terminal where you have installed your node js earlier.
2. then type " npm uninstall -g create-react-app "
3. after that " npm install -g create-react-app"
4. then follow this again [https://tailwindcss.com/docs/guides/create-react-app](url) 
this might help you, actually u may have the "  *npm*  create-react-app " older version then this steps will work.

**Reply by jaikangam** [reply]
Someone still looking for a solution, what @rahulrg1 said is true. Uninstall and get the latest version of create-react-app.

**Reply by Blasto33** [reply]
Thanks a lot, still works fine! The link for the 4th step is now https://v2.tailwindcss.com/docs/guides/create-react-app

**Comment by CoderJakaria** [comment]
Can anyone help me? my client want to go react-scripts@4.0.3 but after going the tailwindcss not not working

**Reply by slimjimrim** [reply]
BUMP - i am having the same problem

**Reply by glasscloud** [reply]
I ran into the same problem today. Needed to downgrade react-scripts from 5.0.1 to 4.0.3 and after doing that tailwind styling no longer worked.

**Reply by melhassouni** [reply]
i had the same issue just today i had to downgrade my react-scripts from 5.0.1 to 4.0.3 but tailwind stopped working. does anyone knows any solution ?

**Comment by jaikangam** [comment]
```npm install -D tailwindcss postcss@latest autoprefixer ``` this works.  But recommended to uninstall the create-react-app and get the latest version. 

I got the same issue. Seem it is because of dependency. 

If you've previously installed ```create-react-app``` globally via ```npm install -g create-react-app```, we recommend you uninstall the package using ```npm uninstall -g create-react-app``` or ```yarn global remove create-react-app``` to ensure that npx always uses the latest version.

This is from the Create-react-app doc.

**Comment by ZeeshanAhmadKhalil** [comment]
It not working, i have to use 

> npx tailwindcss -i ./src/input.css -o ./dist/output.css --watch

to make it work, is there a way which starts it automatically when I start my react app?

**Reply by Orta12** [reply]
Yes it is not working

**Comment by GayatriDunakhe** [comment]
my tailwide css doesn't show any styling on the app 

My App.js - 

![image](https://user-images.githubusercontent.com/74578232/233761003-0e93c67f-bb15-4a2a-b4a3-ffc9fc71911e.png)

My index.css - 

![image](https://user-images.githubusercontent.com/74578232/233761038-86be28b2-fc95-4bee-b73d-edbe462eafbc.png)

My postcss.config.js - 

![image](https://user-images.githubusercontent.com/74578232/233761031-f7373323-33e3-474d-a187-6e6eb33502a6.png)

**Reply by wongjn** [reply]
Consider checking:
- You have `import`ed your `index.css` somewhere.
- Source code that contains Tailwind classes are covered by the `content` file globs.
- You are running `react-scripts@^5`.

**Reply by GayatriDunakhe** [reply]
Ok i will check it! Thank you

**Comment by Shehbaz-alphai** [comment]
Tailwind not working in create-react-app
my tailwind css doesn't show any styling on the app
